File: urbanairship-core/src/main/java/com/urbanairship/json/matchers/NumberRangeMatcher.java
/* Copyright Airship and Contributors */

package com.urbanairship.json.matchers;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import com.urbanairship.json.JsonMap;
import com.urbanairship.json.JsonValue;
import com.urbanairship.json.ValueMatcher;

/**
 * Range matcher.
 *
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class NumberRangeMatcher extends ValueMatcher {

    @NonNull
    public static final String MIN_VALUE_KEY = "at_least";

    @NonNull
    public static final String MAX_VALUE_KEY = "at_most";

    @Nullable
    private final Double min;

    @Nullable
    private final Double max;

    /**
     * Default constructor.
     *
     * @param min The min value.
     * @param max The max value.
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    public NumberRangeMatcher(@Nullable Double min, @Nullable Double max) {
        this.min = min;
        this.max = max;
    }

    @Override
    public boolean equals(@Nullable Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        NumberRangeMatcher that = (NumberRangeMatcher) o;

        if (min != null ? !min.equals(that.min) : that.min != null) {
            return false;
        }
        return max != null ? max.equals(that.max) : that.max == null;
    }

    @Override
    public int hashCode() {
        int result = min != null ? min.hashCode() : 0;
        result = 31 * result + (max != null ? max.hashCode() : 0);
        return result;
    }

    @Override
    protected boolean apply(@NonNull JsonValue value, boolean ignoreCase) {
        if (min != null && (!value.isNumber() || value.getDouble(0) < min)) {
            return false;
        }

        return max == null || (value.isNumber() && !(value.getDouble(0) > max));
    }

    @NonNull
    @Override
    public JsonValue toJsonValue() {
        return JsonMap.newBuilder()
                      .putOpt(MIN_VALUE_KEY, min)
                      .putOpt(MAX_VALUE_KEY, max)
                      .build()
                      .toJsonValue();
    }

}

File: urbanairship-core/src/main/java/com/urbanairship/json/matchers/NumberRangeMatcher.kt
/* Copyright Airship and Contributors */
package com.urbanairship.json.matchers

import androidx.core.util.ObjectsCompat
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonValue
import com.urbanairship.json.ValueMatcher
import com.urbanairship.json.jsonMapOf

/**
 * Range matcher.
 *
 * @hide
 */
internal class NumberRangeMatcher(
    private val min: Double?,
    private val max: Double?
) : ValueMatcher() {

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }

        val that = other as NumberRangeMatcher

        if (min != that.min) { return false }
        if (max != that.max) { return false }

        return true
    }

    override fun hashCode(): Int {
        return ObjectsCompat.hash(min, max)
    }

    override fun apply(jsonValue: JsonValue, ignoreCase: Boolean): Boolean {
        if (min == null && max == null) {
            return true
        }

        if (!jsonValue.isNumber) {
            return false
        }

        min?.let {
            if (jsonValue.getDouble(0.0) < it) {
                return false
            }
        }

        max?.let {
            if (jsonValue.getDouble(0.0) > it) {
                return false
            }
        }

        return true
    }

    @Throws(JsonException::class)
    override fun toJsonValue(): JsonValue = jsonMapOf(
        MIN_VALUE_KEY to min,
        MAX_VALUE_KEY to max
    ).toJsonValue()

    companion object {
        const val MIN_VALUE_KEY: String = "at_least"
        const val MAX_VALUE_KEY: String = "at_most"
    }
}
